Handle modded packets on the network thread

SquidDev opened the pull request:
Fixes #8642. This ensures custom modded packets are both read and handled on the network thread by default, rather than the main server/client thread.

Tested on both single player and a server, with Netty's Advanced leak detection enabled. Custom packets work as expected (tested with [a mod of mine](https://github.com/cc-tweaked/CC-Tweaked)), and it does not appear to introduce any `ByteBuf` leaks.

I did also look at changing the custom packets to write their contents on the network thread (they are currently written when the packet is created instead), but this would require a more invasive change, and I wasn't comfortable doing that without further discussion.

SquidDev committed:
Handle modded packets on the network thread

 - On the server we simply need to remove the call to
   ensureRunningOnSameThread.

 - On the client side, we now handle the packet at the very start of the
   call. We make sure we're running from a network thread to prevent
   calling the handling code twice.

   While this does mean we no longer call .release(), in practice this
   doesn't cause any leaks as ClientboundCustomPayloadPacket releases
   for us.

sciwhiz12 reviewed (changes requested):
While I know @cpw has commented on the issue that packets being handled on the network thread is the original design, I'd like for an example situation to be provided in which handling the packet on the network thread is more preferrable than handling it on the render/server thread. Other than that, this PR looks fine to me.

---

Please add `fixes #8642` to your PR description so this PR is automatically linked to the issue.

SquidDev commented:
What originally caused me to spot this was that packet decoding is also handled on the main thread. I compress some of the bulkier packets when sending[^1] and the corresponding decompression was showing up in profiles during some stress tests. If that can be shunted off onto the network thread, it does result in moderate FPS improvements.

[^1]: The server-side compression will still be done on the main thread as mentioned in the OP, but that's much harder to fix.

sciwhiz12 commented:
Hmmm, so this would introduce an asymmetry for which thread packets are encoded and decoded -- packets would be encoded on the main thread while decoded on the network thread. Maybe we should document this behavior somewhere (perhaps as javadocs on `SimpleChannel.MessageBuilder#encoder` and `SimpleChannel.MessageBuilder#decoder`?) so developers which somehow rely on this behavior are aware of this.

SquidDev committed:
Clarify behaviour a little in the documentation

sciwhiz12 commented on src/main/java/net/minecraftforge/network/simple/SimpleChannel.java:
```diff
@@ -181,11 +181,34 @@ private static <MSG> MessageBuilder<MSG> forType(final SimpleChannel channel, fi
+        /**
+         * Set the message encoder, which writes this message to a {@link FriendlyByteBuf}.
+         *
+         * The encoder is called <em>immediately</em> {@link #send(PacketDistributor.PacketTarget, Object) when the
+         * packet is sent}. This means encoding typically occurs on the main server/client thread rather than on the
+         * network thread.
+         *
+         * However, this behaviour should not be relied on, and the encoder should try to be thread-safe and not
+         * interact with the current game state.
```
Paragraphs other than the first should either be separated by or surrounded with `<p>` tags. (I personally prefer the latter.) Additionally, `{@linkplain}` should be used when the label would normally not be in monospace.

```suggestion
         * <p>The encoder is called <em>immediately</em> {@linkplain #send(PacketDistributor.PacketTarget, Object) when the
         * packet is sent}. This means encoding typically occurs on the main server/client thread rather than on the
         * network thread.</p>
         *
         * <p>However, this behaviour should not be relied on, and the encoder should try to be thread-safe and not
         * interact with the current game state.</p>
```

SquidDev replied:
> Additionally, `{@linkplain}` should be used when the label would normally not be in monospace.

Oh, I wasn't that existed. Oh, that's very nice.

sciwhiz12 commented on src/main/java/net/minecraftforge/network/simple/SimpleChannel.java:
```diff
@@ -181,11 +181,34 @@ private static <MSG> MessageBuilder<MSG> forType(final SimpleChannel channel, fi
+        /**
+         * Set the message decoder, which reads the message from a {@link FriendlyByteBuf}.
+         *
+         * The decoder is called when the message is received on the network thread. The decoder should not attempt to
+         * access or mutate any game state, deferring that until the {@link #consumer(ToBooleanBiFunction) the message
+         * is handled}.
```
Paragraphs other than the first should either be separated by or surrounded with `<p>` tags. (I personally prefer the latter.) Additionally, `{@linkplain}` should be used when the label would normally not be in monospace.

```suggestion
         * <p>The decoder is called when the message is received on the network thread. The decoder should not attempt to
         * access or mutate any game state, deferring that until the {@linkplain #consumer(ToBooleanBiFunction) the message
         * is handled}.</p>
```

sciwhiz12 commented on src/main/java/net/minecraftforge/network/simple/SimpleChannel.java:
```diff
@@ -223,6 +246,17 @@ public MessageBuilder<MSG> noResponse()
+         * Set the message consumer, which is called once a message has been decoded. This accepts the decoded message
+         * object and the message's context.
+         *
+         * The consumer is called on the network thread, and so should not interact with most game state by default.
+         * {@link NetworkEvent.Context#enqueueWork(Runnable)} can be used to handle the message on the main server or
+         * client thread.
```
Paragraphs other than the first should either be separated by or surrounded with <p> tags. (I personally prefer the latter.)

```suggestion
         * <p>The consumer is called on the network thread, and so should not interact with most game state by default.
         * {@link NetworkEvent.Context#enqueueWork(Runnable)} can be used to handle the message on the main server or
         * client thread.</p>
```

sciwhiz12 reviewed (changes requested):
Looks fine, just want to fixup the formatting of the javadocs. Also, run the `:forge:javadoc` task to see if the `javadoc` tool complains about the new javadocs.

SquidDev committed:
Javadoc formatting

SquidDev commented:
`./gradlew :forgeJavadoc` seems fine. It gives a single error, but appears unrelated to this change.

```
> Task :forge:javadoc
/home/squid/Documents/software/MinecraftForge/src/main/java/net/minecraftforge/common/world/ModifiableBiomeInfo.java:114: error: @param name not found
             * @param biome A biome that has not had its modified biome data generated yet
                      ^
1 error
```

SquidDev committed:
Add a helper method for handling packets on the main thread

Also rename the network thread one. Should make it clearer the expected
behaviour of the two, and make it clearer there's a potentially breaking
change.

cpw commented on src/main/java/net/minecraftforge/network/simple/SimpleChannel.java:
```diff
@@ -223,11 +246,44 @@ public MessageBuilder<MSG> noResponse()
             return this;
         }
 
-        public MessageBuilder<MSG> consumer(BiConsumer<MSG, Supplier<NetworkEvent.Context>> consumer) {
```
There's no need to binary break here. Mark consumer as @Deprecated and redirect to consumerMainThread in it. Otherwise this looks fine. I need to look at encode/decode offthread too.

SquidDev committed:
Add back consumer() methods

Also document EventNetworkChannel, to clarify the thread behaviour
there.

SquidDev committed:
Add since = "1.19" to deprecated annotations

SquidDev commented:
Thanks! ❤️

KingLemming commented:
So this appears to have completely busted my networking. I think there's a bug introduced here beyond the scope of just fixing a perceived issue.

SquidDev commented:
@KingLemming Can you clarify what you mean by "busted your networking"? Is there a bug report/reproduction case I can look at?

gigaherz commented:
We have troubleshooted on one of the internal channels. Was simply a misplaced call to read() from the packet. It didn't break before this PR because of enqueueWork running immediately.

KingLemming commented:
Can safely ignore, as ghz said.
